src/mrinufft/trajectories/io.py

```diff
@@ -0,0 +1,377 @@
+            left_over -= 1
+        file.write(str('0\n'*left_over))
+    # Write all the k0 values
+    file.write('\n'.join(
+        ' '.join(
+                ["{0:5.4f}".format(iter2) for iter2 in iter1]
+            )
+        for iter1 in start_positions) + '\n'
+    )
+    if version < 4.1:
+        # Write the maximum Gradient
+        file.write(str(max_grad) + '\n')
+    # Normalize gradients
+    gradients = gradients / max_grad
+    file.write('\n'.join(
+        ' '.join(
+            ["{0:5.6f}".format(iter2) for iter2 in iter1]
+        )
+        for iter1 in gradients) + '\n'
+    )
+    file.close()
+    y = []
+    with open(grad_filename + '.txt', 'r') as txtfile:
+        for line in txtfile:
+            x = line.split(' ')
+            for val in x:
+                y.append(float(val))
+    float_array = array('f', y)
+    with open(grad_filename + '.bin', 'wb') as binfile:
+        float_array.tofile(binfile)
+    if not keep_txt_file:
+        os.remove(grad_filename + '.txt')
+
+
+
+def _pop_elements(array, num_elements=1, type='float'):
+    """A function to pop elements from an array.
+    Parameters
+    ----------
+    array : np.ndarray
+        Array to pop elements from.
+    num_elements : int, optional
+        number of elements to pop, by default 1
+    type : str, optional
+        Type of the element being popped, by default 'float'
+    
+    Returns
+    -------
+    element_popped: 
+        Element popped from array with type as specified.
+    array: np.ndarray
+        Array with elements popped.
+    """
+    
+    if num_elements == 1:
+        return array[0].astype(type), array[1:]
+    else:
+        return array[0:num_elements].astype(type), \
+               array[num_elements:]
+                   
+                   
+def get_kspace_loc_from_gradfile(grad_filename: str, dwell_time: float=0.01, num_adc_samples: int=None, 
+                                 gyromagnetic_constant: float=42.576e3, gradient_raster_time: float=0.010,
+                                 read_shots: bool=False):
+    """Get k-space locations from gradient file.
+
+    Parameters
+    ----------
+    grad_filename : str
+        Gradient filename.
+    dwell_time : float, optional
+        Dwell time of ADC, by default 0.01
+    num_adc_samples : int, optional
+        Number of ADC samples, by default None
+    gyromagnetic_constant : float, optional
+        Gyromagnetic Constant, by default 42.576e3
+    gradient_raster_time : float, optional
+        Gradient raster time, by default 0.010
+    read_shots : bool, optional
+        Whether in read shots configuration which accepts an extra point at end, by default False
+
+    Returns
+    -------
+    kspace_loc : np.ndarray
+        K-space locations. Shape (num_shots, num_adc_samples, dimension).
+    """    
+    dwell_time_ns = dwell_time * 1e6
+    gradient_raster_time_ns = gradient_raster_time * 1e6
+    with open(grad_filename, 'rb') as binfile:
```

Comment:
just open the file and close it at the end, winning an indentation level.

Reply:
xD I thought this was pythonic, I can do that no issues, but then we need to handle errors?